apps/slippirank/slippirank.star

```diff
@@ -0,0 +1,261 @@
+}
+
+def getRank(elo):
+    for rank in RANKS:
+        if rank["min"] < elo and rank["max"] > elo:
+            return rank["name"]
+    return "Unranked3"
+
+RANK_URL = "https://gql-gateway-dot-slippi.uc.r.appspot.com/graphql"
+
+def getUserCodeDashIndex(userCode):
+    for i in range(len(userCode)):
+        if userCode[i] == "#" or userCode[i] == "-":
+            return i
+    return -1
+
+def requestRank(userCode):
+    body = json.encode({
+        "operationName": "AccountManagementPageQuery",
+        "variables": {
+            "cc": userCode,
+            "uid": userCode,
+        },
+        "query": "fragment userProfilePage on User {\n  fbUid\n  displayName\n  connectCode {\n    code\n    __typename\n  }\n  status\n  activeSubscription {\n    level\n    hasGiftSub\n    __typename\n  }\n  rankedNetplayProfile {\n    id\n    ratingOrdinal\n    ratingUpdateCount\n    wins\n    losses\n    dailyGlobalPlacement\n    dailyRegionalPlacement\n    continent\n    characters {\n      id\n      character\n      gameCount\n      __typename\n    }\n    __typename\n  }\n  __typename\n}\n\nquery AccountManagementPageQuery($cc: String!, $uid: String!) {\n  getUser(fbUid: $uid) {\n    ...userProfilePage\n    __typename\n  }\n  getConnectCode(code: $cc) {\n    user {\n      ...userProfilePage\n      __typename\n    }\n    __typename\n  }\n}\n",
+    })
+    res = http.post(
+        RANK_URL,
+        body = body,
+        headers = {
+            "Content-Type": "application/json",
+        },
+    )
+    if res.status_code != 200:
+        fail("request failed with status %d", res.status_code)
+    res = res.json()
+    return res
+
+REFRESH_TIME = 43200  # twice a day
+DEFAULT_USER_CODE = "hbox-305"
+
+def main(config):
+    userCode = config.str("userCode")
+    showRankName = config.bool("showRankName")
+    showElo = config.bool("showElo")
+    if userCode == None or userCode == "":
+        userCode = DEFAULT_USER_CODE
+        # fail("No user code configured")
+
+    userCode = userCode.upper()
+    userCodeDashIndex = getUserCodeDashIndex(userCode)
+    if userCodeDashIndex == -1:
+        fail("Invalid user code")
+
+    # print(userCode)
+
+    userCodeHash = userCode[:userCodeDashIndex] + "#" + userCode[userCodeDashIndex + 1:]
+    rankedData = cache.get("rankedData")
```

Comment:
The cache module is scoped to the app, but not to individual users of the app, or individual installations of the app. In this case, the `rankedData` key will be used by all users, regardless of the configured `userCode`.

The simplest, and best fix would likely be to switch to using the built in cache functionality of the http module. Simply pass `ttl_seconds` to `http.post()` and the response will be cached accordingly.